programs/java_lookup.py

```diff
@@ -22,70 +22,79 @@
-JDK_PATH_REGEXES = [
-    re.compile(
+JDK_PATH_REGEXES = (
+    [re.compile(
         r"^/Library/Java/JavaVirtualMachines/"
-        + JDK_8_AND_UNDER_PATH_VERSION_REGEX_STRING
+        + r
         + "/Contents/Home/*$"
-    ),
-    re.compile(
+    ) for r in JDK_8_AND_UNDER_PATH_VERSION_REGEX_STRINGS] +
+    [re.compile(
         r"^/Library/Java/JavaVirtualMachines/"
-        + JDK_9_AND_OVER_PATH_VERSION_REGEX_STRING
+        + r
         + "/Contents/Home/*$"
-    ),
-    re.compile(
+    ) for r in JDK_9_AND_OVER_PATH_VERSION_REGEX_STRINGS] +
+    [re.compile(
         "^C:\\\\Program Files\\\\Java\\\\"
-        + JDK_8_AND_UNDER_PATH_VERSION_REGEX_STRING
+        + r
         + "(\\\\)*$"
-    ),
-    re.compile(
+    ) for r in JDK_8_AND_UNDER_PATH_VERSION_REGEX_STRINGS] +
+    [re.compile(
         "^C:\\\\Program Files\\\\Java\\\\"
-        + JDK_9_AND_OVER_PATH_VERSION_REGEX_STRING
+        + r
         + "(\\\\)*$"
-    ),
-]
+    ) for r in JDK_9_AND_OVER_PATH_VERSION_REGEX_STRINGS]
+)
 
 
 def _get_suspected_java_version_from_java_path(java_path):
     for regex in JDK_PATH_REGEXES:
         match = regex.match(java_path)
         if match:
-            return int(match.group(2))
+            return int(match.group('major'))
     return None
 
 
 def _get_java_path_for_highest_minor_version(base_path, desired_major_version):
     max_version = None
     max_dir = None
+    regexes = (
+        JDK_8_AND_UNDER_PATH_VERSION_REGEXES
+        if desired_major_version <= 8
+        else JDK_9_AND_OVER_PATH_VERSION_REGEXES
+    )
     for _, dirs, _ in os.walk(base_path):
         for dir in dirs:
-            regex = (
-                JDK_8_AND_UNDER_PATH_VERSION_REGEX
-                if desired_major_version <= 8
-                else JDK_9_AND_OVER_PATH_VERSION_REGEX
-            )
-            match = regex.match(dir)
-            if match:
-                major_version = int(match.group(2))
-                if major_version == desired_major_version:
-                    version_string = match.group(1)
-                    version = tuple(
-                        map(int, version_string.replace("_", ".").split("."))
-                    )
-                    if not max_version or version > max_version:
-                        max_version = version
-                        max_dir = dir
+            for regex in regexes:
+                match = regex.match(dir)
+                if match:
+                    major_version = int(match.group('major'))
+                    if major_version == desired_major_version:
+                        version_string = match.group('full')
+                        version = tuple(
+                            map(int, version_string.replace("_", ".").split("."))
+                        )
+                        if not max_version or version > max_version:
+                            max_version = version
+                            max_dir = dir
```

Comment:
Only concern here is that dirs from os.walk is non-deterministic afaik. I'm pretty sure it uses readdir under the hood. We should probably sort that list, and also ensure that we pick one of openjdk or oracle deterministically if both are present. The determinism problem was already present, but is probably more of an issue now, as it's not horribly unlikely that people have two jdks installed.

Reply:
Updated. I also removed the usage of `os.walk` which was technically wrong for this use case. Since it recursively reads the directory tree, it could have seen a subdirectory of the JDK and then tried to use that directory name as JAVA_HOME without including the parent path - see `return os.path.join(base_path, max_dir) if max_dir else None`.